### ISSUE TITLE
Can StringByteSize in demangleStringLiteral in MicrosoftDemangle.cpp overflow?

### ISSUE THREAD

**Issue by shafik** [llvm/llvm-project#129970]
A static analysis tool flagged this line for possible overflow:

https://github.com/llvm/llvm-project/blob/213028556419cb734be5aadcfd9798b40a8ebc31/llvm/lib/Demangle/MicrosoftDemangle.cpp#L1382

as a possible overflow risk. I can see that we used to check but switched to checking `MangledName.size() < 2` w/ this commit: a0ac65c98f281

The relationship between `MangledName.size()` and `StringByteSize` does not look like it is verified, although the assumption seems to be that it will at least as large as `MangledName.size()`.

The question is whether `StringByteSize` could ever have a value of `1` or `0` before decrementing by `2` and thereflow overflowing to max unsigned value or max unsigned value - 1.

**Comment by shafik** [comment]
CC @nico who was author of the commit mentioned above.

**Comment by shafik** [comment]
ping @nico I would really appreciate your eyes on this so I can understand your changes a bit better and hopefully that helps to identify if this is a false positive or not.

**Comment by shafik** [comment]
ping @nico

**Comment by tahonermann** [comment]
I spent quite a while looking at the code and eventually concluded that the static analysis report highlights a valid issue. Consider the following code.
```
inline const wchar_t* f() {
  return L"tex";
}
auto v = f();
```
The string literal returned by the inline function requires a symbol. The mangled name produced by MSVC is `??_C@_17EEHFKJGG@?$AAt?$AAe?$AAx@`. The portion of the name that corresponds to the part parsed by `demangleStringLiteral()` is `@_17EEHFKJGG@?$AAt?$AAe?$AAx@` and it breaks down as follows:
```
@_ 1 7 EEHFKJGG @ ?$AAt ?$AAe ?$AAx @
|  | | |        | |     |     |     '- end marker
|  | | |        | |     |     '------- encoded character (L'x')
|  | | |        | |     '------------- encoded character (L'e')
|  | | |        | '------------------- encoded character (L't')
|  | | |        '--------------------- CRC separator
|  | | '------------------------------ CRC
|  | '-------------------------------- encoded byte length (8, including implicit null terminator)
|  '---------------------------------- wide string marker (1==wchar_t)
'------------------------------------- string literal marker
```

The code currently lacks a check for the encoded byte length (corresponding to `StringByteSize`) to ensure it is an even number for a wide string. The question is what happens if the encoded length is off by one (or more!) relative to the number of encoded characters. At present, no error is detected. I think all of the following demangling attempts should fail.
```
$ llvm-undname
# With encoded length 3.
??_C@_12EEHFKJGG@?$AAt?$AAe?$AAx@
L"tex"

# With encoded length 7.
??_C@_16EEHFKJGG@?$AAt?$AAe?$AAx@
L"tex"

# With encoded length 9.
??_C@_18EEHFKJGG@?$AAt?$AAe?$AAx@
L"tex"
```

I think the following change would make the code easier to read and understand. Don't mutate the value of `StringByteSize` after it is read. Instead, track the number of character literals that are decoded and compare that computed result with `StringByteSize` (with adjustment for wide characters encoding two bytes, the null terminator not being explicitly encoded, and that a maximum of 32 characters are encoded (in which case, set `IsTruncated` accordingly)). Do this for both the `char` (and `charN_t`) and `wchar_t` cases.
